core/src/main/scala/com/microsoft/azure/synapse/ml/causal/SyntheticEstimator.scala

```diff
@@ -0,0 +1,234 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.azure.synapse.ml.causal
+
+import org.apache.spark.sql.expressions.{UserDefinedFunction, Window}
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.{DataFrame, Encoder, Encoders, Row}
+import org.apache.spark.sql.types._
+import breeze.linalg.{DenseMatrix => BDM, DenseVector => BDV}
+import com.microsoft.azure.synapse.ml.causal.linalg._
+import com.microsoft.azure.synapse.ml.causal.opt.ConstrainedLeastSquare
+import com.microsoft.azure.synapse.ml.logging.SynapseMLLogging
+
+trait SyntheticEstimator extends SynapseMLLogging {
+  this: BaseDiffInDiffEstimator with SyntheticEstimatorParams =>
+
+  import SyntheticEstimator._
+
+  implicit val vectorOps: DVectorOps.type = DVectorOps
+  implicit val matrixOps: DMatrixOps.type = DMatrixOps
+  implicit val matrixEntryEncoder: Encoder[MatrixEntry] = Encoders.product[MatrixEntry]
+  implicit val vectorEntryEncoder: Encoder[VectorEntry] = Encoders.product[VectorEntry]
+
+  private[causal] lazy val postTreatment = col(getPostTreatmentCol)
+  private[causal] lazy val treatment = col(getTreatmentCol)
+  private[causal] lazy val outcome = col(getOutcomeCol)
+  private[causal] val weightsCol = "weights"
+  private[causal] val epsilon = 1E-10
```

Comment:
might want to make settable

Reply:
The weightsCol is entirely internal, and it's never exposed to the caller. I changed it to use findUnusedColumn though. The epsilon is changed to a param now.